src/bootstrap/compile.rs

```diff
@@ -817,6 +817,12 @@ impl Step for Rustc {
         if compiler.stage != 0 {
             match builder.config.rust_lto {
                 RustcLto::Thin | RustcLto::Fat => {
+                    if target.contains("windows") {
```

Comment:
(we know windows-msvc has the miscompile but it could be fine on windows-gnu.)

Reply:
Good point. I suppose that we only definitely know about miscompiles in `x86_64-pc-windows-msvc`, right? So maybe we should just warn against this single target.

Reply:
Yeah, for sure on `x86_64-pc-windows-msvc`, on the others there have been no reports yet I don't think, but it may just be that they're less commonly used. But now that we know a few ways to trigger them, it may be possible to test there.

Reply:
Hm, do the current reports affect out-of-the-box msvc or just when using lld?

Reply:
lld is usually involved (e.g. #103353), and it should also be the case in #109114

Reply:
Right so we don't currently build with lld in CI. Maybe we should but isn't that effectively a new target, in a similar way to we have both `pc-windows-gnu`  and `pc-windows-gnullvm`? It's maybe a bit heavy handed to disable thinlto for `x86_64-pc-windows-msvc` when the issue is with an ambiguously supported configuration.

Reply:
@lqd Are you able to reproduce the miscompilation locally? If yes, could you please try if it also happens without LLD?

Reply:
#109114 (if we don't use lld on CI) coupled with #113425 (which involves lld) together mean that building rustc with `rust.lto = "thin"` will result in a miscompiled rustc. 

We're not building rustc with ThinLTO on windows because of these issues, and those `-msvc` miscompilations look serious enough that people probably shouldn't do so either, and instead be made aware of that? 

So I'm not sure what you mean by "maybe a bit heavy handed to disable thinlto for `x86_64-pc-windows-msvc`" ? This PR is not disabling ThinLTO, it's about bootstrap preventing building rustc with `-Zdylib-lto`, to avoid miscompilations.

Reply:
We can (hopefully) test for issues like #109114 in CI. But the same is not true for lld, no?

Reply:
https://github.com/rust-lang/rust/issues/109114 => Rustc compiled with LLD=false, ThinLTO=true => produces miscompilations
https://github.com/rust-lang/rust/pull/112946#issuecomment-1622565305 => Rustc compiled with LLD=true, ThinLTO=true => produces miscompilations

So it seems that when we compile rustc with ThinLTO, it then produces miscompilations regardless whether it was itself compiled with LLD or not. Therefore we want to ban ThinLTO for the moment.

Note 1: This has nothing to do with applying LTO to Rust programs. Rustc can still perform LTO when building Rust code on Windows.
Note 2: Rustc on Windows is currently not optimized with LTO because of the miscompilations. This PR doesn't change that.

Reply:
I mean we could also emit a scary warning instead, if an error would disturb your use-cases.

Reply:
FWIW I recall seeing many MSVC + LTO + LLD miscompilation issues over the years but strangely they didn't reproduce on `windows-gnu` when I tested them, so I'd agree it makes sense to apply this limit on `windows-msvc` only.